pkg/drift/assets/assets.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package assets
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
+	asset "cloud.google.com/go/asset/apiv1"
+	"cloud.google.com/go/asset/apiv1/assetpb"
+	"google.golang.org/api/iterator"
+	fmpb "google.golang.org/protobuf/types/known/fieldmaskpb"
+)
+
+const (
+	// Organization Node Type.
+	Organization = "Organization"
+	// Folder Node Type.
+	Folder = "Folder"
+	// Project Node Type.
+	Project = "Project"
+
+	OrganizationAssetType = "cloudresourcemanager.googleapis.com/Organization"
+	FolderAssetType       = "cloudresourcemanager.googleapis.com/Folder"
+	ProjectAssetType      = "cloudresourcemanager.googleapis.com/Project"
+	BucketAssetType       = "storage.googleapis.com/Bucket"
+)
+
+// resourceNameIDPattern is a Regex pattern used to parse ID from ParentFullResourceName.
+var resourceNameIDPattern = regexp.MustCompile(`\/\/cloudresourcemanager\.googleapis\.com\/(?:folders|organizations)\/(\d*)`)
+
+// HierarchyNode represents a node in the GCP Resource Hierarchy.
+// Example: Organization, Folder, or Project.
+type HierarchyNode struct {
+	// The unique identifier of the GCP Organization, Folder, or Project
+	// Example: 123123423423
+	ID string
+	// The unique string name of the Organization, Folder, or Project.
+	// Example: my-project-1234
+	Name string
+	// The unique identifier of the Folder or Organization this Folder or Project resides in.
+	ParentID string
+	// The type of the parent node. Either Folder or Organization.
+	ParentType string
+	// The type of node. Either Folder or Organization or Project
+	NodeType string
+}
```

Comment:
Now that I see the comments, I think this struct might benefit from an interface instead of concrete types for Project, Folder, and Organization:

```go
type HierarchyNode interface {
  // Number is the numeric identifier for the node.
  // Example: 1234567890
  Number() string

  // Name is the human name of the node.
  Name() string

  // Parent is the next node up the hierarchy.
  Parent() HierarchyNode
}

type ProjectNode struct {
  //...
}

type FolderNode struct {
  //...
}

type OrganizationNode struct {
  //...
}
```

Getting the "type" of the current node or parent node can happen via `(.type)`.

Reply:
I'm leaning away from this as I suspect we will expand this to support a complex hierarchy of other GCP resources that also support IAM (WIF, Cloud Run, etc). I'd rather not have to add a concrete type for every new kind of resource that gets supported.